scipy/stats/morestats.py

```diff
@@ -2152,17 +2152,90 @@ def anderson_ksamp(samples, midrank=True):
 AnsariResult = namedtuple('AnsariResult', ('statistic', 'pvalue'))
 
 
-def ansari(x, y):
+class _ABW:
+    """Distribution of Ansari-Bradley W-statistic under the null hypothesis."""
+    # TODO: calculate exact distribution considering ties
+    # We could avoid summing over more than half the frequencies,
+    # but inititally it doesn't seem worth the extra complexity
+
+    def __init__(self):
+        """Minimal initializer."""
+        self.m = None
+        self.n = None
+        self.astart = None
+        self.total = None
+        self.freqs = None
+
+    def _recalc(self, n, m):
+        """When necessary, recalculate exact distribution."""
+        if n != self.n or m != self.m:
+            self.n, self.m = n, m
+            # distribution is NOT symmetric when m + n is odd
+            # n is len(x), m is len(y), and ratio of scales is defined x/y
+            astart, a1, _ = statlib.gscale(n, m)
+            self.astart = astart  # minimum value of statistic
+            # Exact distribution of test statistic under null hypothesis
+            # expressed as frequencies/counts/integers to maintain precision.
+            # Stored as floats to avoid overflow of sums.
+            self.freqs = a1.astype(np.float64)
+            self.total = self.freqs.sum()  # could calculate from m and n
+            # probability mass is self.freqs / self.total;
+
+    def pmf(self, k, n, m):
+        """Probability mass function."""
+        self._recalc(n, m)
+        # The convention here is that PMF at k = 12.5 is the same as at k = 12,
+        # -> use `floor` in case of ties.
+        ind = np.floor(k - self.astart).astype(int)
+        return self.freqs[ind] / self.total
+
+    def cdf(self, k, n, m):
+        """Cumulative distribution function."""
+        self._recalc(n, m)
+        # This is a bit subtle; easiest to think through an example:
+        # the CDF at k = 12.5 should be the same as at k = 12 -> `floor`
+        # The convention is for CDF at k = 12 to included the probability mass
+        # at k = 12 -> slice `:ind+1`.
+        ind = np.floor(k - self.astart).astype(int)
+        return self.freqs[:ind+1].sum() / self.total
+
+    def sf(self, k, n, m):
+        """Survival function."""
+        self._recalc(n, m)
+        # This is a bit subtle; easiest to think through an example:
+        # the SF at k = 12.5 should be the same at k = 13 -> `ceil`.
+        # The convention is for  SF at k = 13 to included the probability mass
+        # at k = 13 -> slice `ind:`.
+        ind = np.ceil(k - self.astart).astype(int)
+        return self.freqs[ind:].sum() / self.total
```

Comment:
```suggestion
    def cdf(self, k, n, m):
        """Cumulative distribution function."""
        self._recalc(n, m)
        # Null distribution derived without considering ties is
        # approximate. Round down to avoid Type I error.
        ind = np.ceil(k - self.astart).astype(int)
        return self.freqs[:ind+1].sum() / self.total

    def sf(self, k, n, m):
        """Survival function."""
        self._recalc(n, m)
        # Null distribution derived without considering ties is
        # approximate. Round down to avoid Type I error.
        ind = np.floor(k - self.astart).astype(int)
        return self.freqs[ind:].sum() / self.total
```
Note that this change doesn't affect the behavior of the code as-is; we can't get to this code if there are ties, so the direction of rounding doesn't matter. However, Hollander and Wolfe _Nonparametric Statistical Methods_ states that if there are ties, we can look up the critical value of the test statistic for a given confidence level just as we would if there were not ties. Based on that statement:

* It is reasonable to allow the user to specify `method = 'exact'` (as we had in this PR before)
* We should round in this direction (as corrected above) if we want the p-value to be consistent with the statement

I can't find an example to compare against. Ansari.test in R won't do exact p-values with ties, and SAS considers ties when deriving the null distribution, so neither can be used for direct comparison. But I have seen that rounding in the other direction (before the change above) leads to p-values that are too low compared to the exact p values considering ties. I think we'd prefer that the p-value be too high - reduced change of type I error at the expense of increased change of type II error.

I wrote a much longer explanation; let me know if you want to read it. But it would probably be better if others considered this independently.

Again; this doesn't affect the behavior of the code right now. It's only for if this pathway gets opened up with a `method` parameter that lets user's calculate "exact" p-values when there are ties but with a null distribution derived without considering ties. In this case, the p-value would still be approximate, and I think it's better for the approximation to overestimate the p-value.

Reply:
> Based on that statement:
> 
> * It is reasonable to allow the user to specify `method = 'exact'` (as we had in this PR before)
> * We should round in this direction (as corrected above) if we want the p-value to be consistent with the statement

Agreed with both points.

> I can't find an example to compare against. Ansari.test in R won't do exact p-values with ties, and SAS considers ties when deriving the null distribution, so neither can be used for direct comparison.

`exactRankTests`'s [`ansari.exact`](https://rdrr.io/cran/exactRankTests/man/ansari.exact.html) calculates exact p-values using the mid-ranks method when ties are present. Do you think we can use that here? I'll try.

> But I have seen that rounding in the other direction (before the change above) leads to p-values that are too low compared to the exact p values considering ties. I think we'd prefer that the p-value be too high - reduced change of type I error at the expense of increased change of type II error.

This makes sense. I definitely agree that the test should avoid making Type I errors and increased Type II errors doesn't seem like a big sacrifice. But we should probably mention that in the documentation while adding the `method` parameter.

> Again; this doesn't affect the behavior of the code right now. It's only for if this pathway gets opened up with a `method` parameter that lets user's calculate "exact" p-values when there are ties but with a null distribution derived without considering ties. In this case, the p-value would still be approximate, and I think it's better for the approximation to overestimate the p-value.

I agree! I will commit this change and try to resolve the documentation failures. I guess that will check the remaining things out.

> I wrote a much longer explanation; let me know if you want to read it. But it would probably be better if others considered this independently.

Thanks, I would appreciate a longer explaination. In case you think it would get in others' way, I can wait :)

Reply:
> `exactRankTests`'s `ansari.exact` calculates exact p-values using the mid-ranks method when ties are present. Do you think we can use that here?

Hmm depends on what we mean by "use that here". 

If you mean "answer the question", then no. I'm trying to get an outside reference for which way to round the statistic when it is not an integer so that we can compare it against the null distribution derived without considering ties, which has only integer support. Based on Hollander and Wolfe, that is a reasonable thing to do, but I can't find an example. `ansari.exact` produces exact p-values for a null distribution that _does_ consider ties. 

If by "use that here" you mean "calculate exact p-values for with a null distribution that considers ties", then yes, ultimately that's what we'd like to do. gh-13899 will do that for small samples, but it implements only the naive, brute-force approach (slow) and the random permutation approach (approximate). We want something faster and exact, which would probably have to be specific to the `ansari` statistic. If we can figure out what `ansari.exact` (and SAS) do (e.g. maybe it's a recursive approach) without looking at the source code, great. I haven't looked hard for a reference, so it would be great if you'd like to research it.

Reply:
> I would appreciate a longer explanation

OK! You asked for it! Here's my original answer:

If I am reading the original code correctly, the rounding we are doing here in case of ties is the same as what was happening before. For instance, before the refactoring, the survival function was calculated in lines 2179-2183 as:
```
            find = int(floor(ind))
            if ind == floor(ind):
                pval = 2.0 * np.sum(a1[find:], axis=0) / total
            else:
                pval = 2.0 * np.sum(a1[find+1:], axis=0) / total
```
Suppose `ind = 39.5`, then `find = 39`, and because `ind != floor(ind)`, we will sum over the slice `a1[40:]`. All this seems equivalent to just:
```
ind = int(ceil(ind))
pval = 2.0 * np.sum(a1[ind:])
```
which is what we now do in the survival function `_ABW.sf`. This was part of the motivation for the refactoring - to simplify this logic.

Based on further reading, though, I now think we should reverse the direction we round when `ind` is not an integer: I now think that in `cdf` we should `ceil` and in `sf` we should `floor`. It doesn't change the behavior of the `ansari` function at all right now, because if there are ties (the only situation in which `k` would be non-integer) the asymptotic approximation is used. In case of ties, we would never get to the `else` case in the code above, and we won't use the `_ABW` class at all. But it _would_ matter if we were to add the `exact` keyword (again) before we implement the null distribution considering ties.

My position now is based on Hollander and Wolfe Nonparametric Statistical Methods. It doesn't explain things in terms of calculating p-values; we have to think in terms of accepting or rejecting the null hypothesis at a given level of significance based on a critical value of the test statistic.

Suppose we are interested in performing a one-sided "greater" test for data with `len(x)=7` and `len(y) = 9`. According to Table A.6 of H&W, the critical value of the test statistic at the 0.0441 level of significance is `40`, and the critical value of the test statistic at the 0.0694 level of significance is `39`. These levels are just the values of the survival function of the null distribution at `k = 39` and `k=40`:
```
>>> frorm
>>> dist = _ABW()
>>> dist.sf(39, 7, 9)
0.0694055944055944
>>> dist.sf(40, 7, 9)
0.044055944055944055
```

If our threshold for significance is 0.0441, then our critical value of the test statistic is 40, not 39. 

Suppose we have ties such that the observed value of the test statistic is 39.5. H&W says "if there are ties among the N observations, use average ranks to compute [the test statistic] and carry on as [you would if there were no ties]". Without ties, we would compare the observed test statistic against the critical value to decide whether to accept or reject the null hypothesis. 

39.5 does _not_ equal or exceed our critical value of 40, so we should not reject the null hypothesis. *Yet if we were to take the ceil of 39.5 in our survival function (as we currently do), it would produce a p-value of 0.044 < 0.05, suggesting that we _should_ reject the null hypothesis.* To keep the behavior consistent with the recommendation of H&W, then, I think we should take the floor of 39.5 in our survival function, so that we do not reject the null hypothesis in this case.

Reply:
> I'm trying to get an outside reference for which way to round the statistic when it is not an integer so that we can compare it against the null distribution derived without considering ties, which has only integer support.

Oh, OK. I thought you wanted to compare against the exact p-value in presence of ties.

> We want something faster and exact, which would probably have to be specific to the `ansari` statistic. If we can figure out what `ansari.exact` (and SAS) do (e.g. maybe it's a recursive approach) without looking at the source code, great. I haven't looked hard for a reference, so it would be great if you'd like to research it. I haven't looked hard for a reference, so it would be great if you'd like to research it

I wasn't able to find anything specific to `ansari`:( after a bit of searching. But I will keep on trying.

Also, I read your long explanation and it is wonderful (and enlightening, of course)! Sorry for bothering you too much. Looks like you put a lot of time to write that. Thanks!

Reply:
> Oh, OK. I thought you wanted to compare against the exact p-value in presence of ties.

Do you mean "compute the p-value in the presence of ties by comparing the statistic against the null distribution derived without considering ties"? Well, I do, or at least I [still](https://github.com/scipy/scipy/pull/13650#discussion_r598162379) think it's reasonable to do so, based on H&W's statement about how to proceed in the presence of ties. That's what this is about - the statistic may can be non-integer, and the null distribution has only integer support, and the question is which way to round to get a p-value.

Reply:
> do you mean "compute the p-value in the presence of ties by comparing the observed statistic against the null distribution derived without considering ties

No, I meant you wanted to compare the p-value obtained using the rounding method you demonstrated above with p-value obtained using some packages that do consider the presence of ties to compute the exact null distribution (to see if the rounding gives a p-value closer to exact p-value. I think that won't be the case here as we just want to avoid Type I error and not calculate a closer estimate of the exact p-value).

Reply:
Oh, I did do a little of that. And you're right - rounding the way we were was _closer_ to the true p-value (in that particular case) but it was _less_ which is not a good thing, IMO. Rounding the new way was much greater, but I think that's the safer way to go if it's going to be an approximation. Thanks!